### ISSUE TITLE
Hardware (physical) qubit as gate parameter raises unhelpful exception

### ISSUE THREAD

**Issue by jlapeyre** [openqasm/openqasm#442]
### Observed behavior

Parsing this code 
```qasm
OPENQASM 3.0;
include "stdgates.inc";

gate mygate c, $1 {
  h c;
}
```
raises a `QASM3ParsingError` exception after catching a` ParseCancellationException` from ANTLR.
<details>
  <summary>Stack trace</summary>
 
 ```
ParseCancellationException                Traceback (most recent call last)
~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/openqasm3/parser.py in parse(input_, permissive)
     86     try:
---> 87         tree = parser.program()
     88     except (RecognitionException, ParseCancellationException) as exc:

~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/openqasm3/_antlr/_4_11/qasm3Parser.py in program(self)
    639                 self.state = 131
--> 640                 self.statement()
    641                 self.state = 136

~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/openqasm3/_antlr/_4_11/qasm3Parser.py in statement(self)
    983                     self.state = 168
--> 984                     self.gateStatement()
    985                     pass

~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/openqasm3/_antlr/_4_11/qasm3Parser.py in gateStatement(self)
   3138             self.state = 405
-> 3139             self.scope()
   3140         except RecognitionException as re:

~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/openqasm3/_antlr/_4_11/qasm3Parser.py in scope(self)
   1152             self.state = 186
-> 1153             self.match(qasm3Parser.LBRACE)
   1154             self.state = 190

~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/antlr4/Parser.py in match(self, ttype)
    125         else:
--> 126             t = self._errHandler.recoverInline(self)
    127             if self.buildParseTrees and t.tokenIndex==-1:

~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/antlr4/error/ErrorStrategy.py in recoverInline(self, recognizer)
    702     def recoverInline(self, recognizer:Parser):
--> 703         self.recover(recognizer, InputMismatchException(recognizer))
    704 

~/qiskit/.venvs/env-3.11/lib64/python3.11/site-packages/antlr4/error/ErrorStrategy.py in recover(self, recognizer, e)
    696             context = context.parentCtx
--> 697         raise ParseCancellationException(e)
    698 

ParseCancellationException: None

The above exception was the direct cause of the following exception:

QASM3ParsingError                         Traceback (most recent call last)
```
</details>

### Desired behavior

Raise a more helpful exception, or else accept the code so that it can be validated at a later stage.

### Versions
openqasm 7c98e1b2e8
Python 3.11.0

**Comment by jakelishman** [comment]
This is going to be true of any grammatically invalid program, not just this case. It's not easy to build a parser that gives good error messages, and here ANTLR can't build the parse tree.

Right now, I wouldn't want to devote manpower to trying to add that functionality in. Doing it _right_ is a really big job, and it's kind of a secondary priority; first we could do with having all the happy-path things succeed, before we spend time improving the UX for failures.

(It's also easier to be more lenient in parsing in order to recover/produce good error messages with handwritten parsers, and this package is _meant_ to use the reference ANTLR grammar.)

**Comment by jakelishman** [comment]
Though just to be clear: I do think we could and should do better on error reporting in the long run.

**Comment by jlapeyre** [comment]
I agree that applying a bandaid is not the correct response. I mainly opened this issue to document a specific undesirable feature, and as you point out, the more general issue. Furthermore, I agree that the cost/benefit ratio of implementing something robust to get better diagnostics is rather high at the moment.

One thing that might incidentally take care of this problem is to recognize `\$[\d]+` as an ordinary identifier; that is, don't distinguish physical qubits in the ANTLR lexer an parser. But I wouldn't make such a big change just to solve this particular problem. Also IIUC, the main reason for using ANTLR is to document the grammar in a testable way. Changing this to adapt to a practical use of OC3, namely importing into Qiskit, doesn't necessarily serve that goal. It might be better for the the pure-Python importer to use a different parser. In fact given your interest in a hand-made parser, I wonder what you think of this opinion, Jake:
> [It seems a parse tree doesn't help to meaningfully reduce effort to build AST compared to just rolling out a hand-made recursive descent parser](https://github.com/antlr/antlr4/issues/2428#issuecomment-570644773)

**Comment by jakelishman** [comment]
> Also IIUC, the main reason for using ANTLR is to document the grammar in a testable way. Changing this to adapt to a practical use of OC3, namely importing into Qiskit, doesn't necessarily serve that goal.

I agree with this - this package is serving two masters at the moment, and it's part of the reason I've been more on the side of using a different parser for our needs in Qiskit (my quick-and-dirty `qiskit-qasm3-import` notwithstanding).

> One thing that might incidentally take care of this problem is to recognize `\$[\d]+` as an ordinary identifier; that is, don't distinguish physical qubits in the ANTLR lexer an parser.

I think you're correct here, and if/when #417 does merge I'd vote to change the ANTLR grammar to reflect this view.

On the last opinion: I don't think I have the experience needed to have a properly informed opinion on it, but I've definitely started tending that way after having written my own recursive-descent parsers.  I'm most interested in that link to the Terence Parr (ANTLR author) comment that AST generation is a non-goal of ANTLR 4 - I didn't know he'd said that, and it helps me understand more of ANTLR's exposed public API.
